cmd/keymasterd/2fa_bootstrapOTP.go

```diff
@@ -0,0 +1,134 @@
+			return
+		}
+	case "POST":
+		state.logger.Debugf(3, "Got client POST connection")
+		if err := r.ParseForm(); err != nil {
+			state.logger.Println(err)
+			state.writeFailureResponse(w, r, http.StatusBadRequest,
+				"Error parsing form")
+			return
+		}
+	default:
+		state.writeFailureResponse(w, r, http.StatusMethodNotAllowed, "")
+		return
+	}
+	authUser, currentAuthLevel, err := state.checkAuth(w, r, AuthTypeAny)
+	if err != nil {
+		state.logger.Debugf(1, "%v", err)
+		return
+	}
+	w.(*instrumentedwriter.LoggingWriter).SetUsername(authUser)
+	var inputOTP string
+	if val, ok := r.Form["OTP"]; ok {
+		if len(val) > 1 {
+			state.writeFailureResponse(w, r, http.StatusBadRequest,
+				"Just one OTP Value allowed")
+			state.logger.Printf("Login with multiple OTP Values")
+			return
+		}
+		inputOTP = val[0]
+	}
+	requiredOTP := state.userBootstrapOtp(authUser)
+	if requiredOTP == "" {
+		state.writeFailureResponse(w, r, http.StatusPreconditionFailed,
+			"No valid Bootstrap OTP saved")
+		return
+	}
+	if inputOTP != requiredOTP {
+		state.logger.Printf("Invalid Bootstrap OTP value for %s\n",
+			authUser)
+		state.logger.Debugf(0, "  input: \"%s\" required: \"%s\"\n",
+			inputOTP, requiredOTP)
+		state.writeFailureResponse(w, r, http.StatusUnauthorized,
+			"Invalid Bootstrap OTP")
+		return
+	}
+	profile, _, fromCache, err := state.LoadUserProfile(authUser)
+	if err != nil {
+		state.logger.Printf("error parsing err=%s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		return
+	}
+	if fromCache {
+		state.writeFailureResponse(w, r, http.StatusServiceUnavailable,
+			"Working in DB disconnected mode, try again later")
+		return
+	}
+	profile.BootstrapOTP = bootstrapOTPData{}
+	if err := state.SaveUserProfile(authUser, profile); err != nil {
+		state.logger.Printf("error saving profile randr=%s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		return
+	}
+	_, err = state.updateAuthCookieAuthlevel(w, r,
+		currentAuthLevel|AuthTypeBootstrapOTP)
+	if err != nil {
+		logger.Printf("Auth Cookie NOT found ? %s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError,
+			"Failure when validating Boostrap OTP")
+		return
+	}
+	// eventNotifier.PublishBootstrapOtpAuthEvent(eventmon.AuthTypeBootstrapOTP,
+	// authUser)
+	// Now we send the user to the appropriate place
+	returnAcceptType := getPreferredAcceptType(r)
+	// TODO: The cert backend should depend also on per user preferences.
+	loginResponse := proto.LoginResponse{Message: "success"}
+	switch returnAcceptType {
+	case "text/html":
+		loginDestination := getLoginDestination(r)
+		eventNotifier.PublishWebLoginEvent(authUser)
+		state.logger.Debugf(0, "redirecting to: %s\n", loginDestination)
+		http.Redirect(w, r, loginDestination, 302)
+	default:
+		w.WriteHeader(http.StatusOK)
+		json.NewEncoder(w).Encode(loginResponse)
+	}
+}
+
+func (state *RuntimeState) userBootstrapOtp(username string) string {
```

Comment:
change this to passing in a user profile, and if the profile is cached instead of a username. Getting a user profile means going to the DB and we want to avoid doing that.

Reply:
Done.